ouroboros-consensus/src/Ouroboros/Consensus/Protocol/TPraos.hs

```diff
@@ -0,0 +1,364 @@
+forgeTPraosFields TPraosNodeConfig{..}  TPraosProof{..} mkToSign = do
+    let icn@TPraosIsCoreNode{..} = tpraosIsCoreNode
+        (DiscVKey issuerVK) = ocertVkCold tpraosIsCoreNodeOpCert
+        signedFields = TPraosToSign {
+          tptsIssuerVK = issuerVK
+        , tptsVrfVk = deriveVerKeyVRF tpraosSignKeyVRF
+        , tptsEta = tpraosEta
+        , tptsLeader = tpraosLeader
+        , tptsOCert = tpraosIsCoreNodeOpCert
+        }
+    m <- signedKES
+           (fromIntegral (unSlotNo tpraosProofSlot))
+           (mkToSign signedFields)
+           tpraosIsCoreNodeSKSHot
+    case m of
+      Nothing                  -> error "mkOutoborosPayload: signedKES failed"
+      Just (signature, newKey) -> do
+        putNodeState . Just $ icn { tpraosIsCoreNodeSKSHot = newKey }
+        return $ TPraosFields {
+            tpraosSignature    = signature
+          , tpraosToSign       = signedFields
+          }
+
+{-------------------------------------------------------------------------------
+  TPraos specific types
+-------------------------------------------------------------------------------}
+
+-- | Assembled proof that the issuer has the right to issue a block in the
+-- selected slot.
+data TPraosProof c
+  = TPraosProof
+    { tpraosEta       :: CertifiedVRF (TPraosVRF c) Nonce
+    , tpraosLeader    :: CertifiedVRF (TPraosVRF c) UnitInterval
+    , tpraosProofSlot :: SlotNo
+    , tpraosIsCoreNode :: TPraosIsCoreNode c
+    } deriving Generic
+
+instance TPraosCrypto c => NoUnexpectedThunks (TPraosProof c)
+
+data TPraosLedgerView c = TPraosLedgerView {
+    -- | Stake distribution
+    tpraosLedgerViewPoolDistr :: PoolDistr (TPraosHash c) (TPraosDSIGN c) (TPraosVRF c)
+  , tpraosLedgerViewProtParams :: PParams
+  , tpraosLedgerViewDelegationMap :: Dms (TPraosHash c) (TPraosDSIGN c)
+  , tpraosLedgerViewEpochNonce :: Nonce
+  , tpraosLedgerViewLeaderVal :: UnitInterval
+    -- | Determines which slots are considered to be part of the overlay
+    -- schedule - that is, slots for whom block issuance is reserved to the core
+    -- nodes operating under Ouborobos BFT rules. There are three cases:
+    --
+    -- - The slot is not in the overlay schedule. Then it is considered under
+    --   Praos rules.
+    -- - The slot is reserved in the schedule to a specific genesis keyholder
+    --   ('Just keyHash'). Then the delegate of this stakeholder has the unique
+    --   right to issue a block in this slot.
+    -- - The slot is reserved in the schedule to 'Nothing'. Then nobody is
+    --   eligible to issue a block in this slot.
+    --
+    -- The last situation is required to reconcile the presence of BFT (which
+    -- issues blocks in every slot) in the overlay schedule with the shorter
+    -- slot lengths to be used under Praos, which issues blocks only in some
+    -- proportion (the 'active slot coefficient') of slots.
+  , tpraosLedgerViewOverlaySchedule :: Map.Map Slot (Maybe (GenKeyHash (TPraosHash c) (TPraosDSIGN c)))
```

Comment:
Maybe containing thunks

Reply:
Also - be careful not to duplicate this too much when updating other snapshots

Reply:
For the hard fork combinator: this overlay schedule has a well-defined initial value (which would be derived from `pbftDelegates`).